tests/Microsoft.Identity.Test.Integration.net45/Infrastructure/SeleniumExtensions.cs

```diff
@@ -68,88 +68,100 @@ public static void SaveScreenshot(this IWebDriver driver, TestContext testContex
-
-            if (user.HomeUPN.Contains("outlook.com"))
-            {
-                Trace.WriteLine("Loggin in ... clicking accept promps for outlook.com MSA user");
-                driver.WaitForElementToBeVisibleAndEnabled(By.Id(CoreUiTestConstants.ConsentAcceptId)).Click();
+        public static By ByIds(params string[] ids)
+        {
+            Trace.WriteLine("Finding first elements by id: " + string.Join(" ", ids));
+            string xPathSelector = string.Join(
+                " or ",
+                ids.Select(id => $"@id='{id}'"));
+
+            return By.XPath($".//*[{xPathSelector}]");
+        }
+
+        public static IWebElement FindElementById(this IWebDriver driver, string id)
+        {
+            Trace.WriteLine("Finding element by id: " + id);
+
+            return driver.FindElement(By.Id(id));
+        }
+
+        public static void PerformLogin(this IWebDriver driver, LabUser user, Prompt prompt, bool withLoginHint = false, bool adfsOnly = false)
+        {
+            UserInformationFieldIds fields = new UserInformationFieldIds(user);
+
+            if (adfsOnly && !withLoginHint)
+            {
+                Trace.WriteLine("Logging in ... Entering username");
+                driver.FindElement(By.Id(CoreUiTestConstants.AdfsV4UsernameInputdId)).SendKeys(user.Upn);
             }
+            else
+            {
+                if (!withLoginHint)
+                {
+                    Trace.WriteLine("Logging in ... Entering username");
+                    driver.FindElementById(fields.AADUsernameInputId).SendKeys(user.Upn.Contains("EXT") ? user.HomeUPN : user.Upn);
 
-            if (prompt == Prompt.Consent)
-            {
-                Trace.WriteLine("Consenting...");
-                driver.WaitForElementToBeVisibleAndEnabled(By.Id(fields.AADSignInButtonId)).Click();
-            }
-        }
+                    Trace.WriteLine("Logging in ... Clicking <Next> after username");
+                    driver.FindElementById(fields.AADSignInButtonId).Click();
+                }
+
+                if (user.FederationProvider == FederationProvider.AdfsV2 && user.IsFederated)
+                {
+                    Trace.WriteLine("Logging in ... AFDSv2 - Entering the username again, this time in the ADFSv2 form");
+                    driver.FindElementById(CoreUiTestConstants.AdfsV2WebUsernameInputId).SendKeys(user.Upn);
+                }
+            }
+
+            Trace.WriteLine("Logging in ... Entering password");
+            driver.WaitForElementToBeVisibleAndEnabled(fields.GetPasswordInputId()).SendKeys(user.GetOrFetchPassword());
+
+            Trace.WriteLine("Logging in ... Clicking next after password");
+            driver.WaitForElementToBeVisibleAndEnabled(By.Id(fields.GetPasswordSignInButtonId())).Click();
+
+            if (user.HomeUPN.Contains("outlook.com"))
+            {
+                Trace.WriteLine("Logging in ... clicking accept prompts for outlook.com MSA user");
```

Comment:
>outlook.com MSA user [](start = 76, length = 20)

can you use the actual value here, like user.HomeUPN or something?

Reply:
I think my VS is playing tricks and reformats the line endings. I haven't actually changed anything in this logic.